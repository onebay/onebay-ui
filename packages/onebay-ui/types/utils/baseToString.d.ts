export default function baseToString(value: unknown): string;
