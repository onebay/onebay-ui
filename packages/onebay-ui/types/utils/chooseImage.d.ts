/**
 * 从本地相册选择图片或使用相机拍照。
 * @param {Object} object 参数
 * @param {string[]} [object.sourceType=['album', 'camera']] 选择图片的来源（h5端未实现）
 * @param {string[]} [object.sizeType=['original', 'compressed']] 所选的图片的尺寸（h5端未实现）
 * @param {number} [object.count=9] 最多可以选择的图片张数
 * @param {function} [object.success] 接口调用成功的回调函数
 * @param {function} [object.fail] 接口调用失败的回调函数
 * @param {function} [object.complete] 接口调用结束的回调函数（调用成功、失败都会执行）
 * @param {string} [object.imageId] 用来上传的input元素ID（仅h5端）
 */
export interface TempFile {
    path: string;
    size: number;
    type: string;
}
export interface ChooseImageResponse {
    errMsg: string;
    tempFilePaths: string[];
    tempFiles: TempFile[];
}
export interface ChooseImageOptions {
    count?: number;
    success?: (resp: ChooseImageResponse) => unknown;
    fail?: (resp: ChooseImageResponse) => unknown;
    complete?: (resp: ChooseImageResponse) => unknown;
    imageId?: string;
}
declare const chooseImage: (options: ChooseImageOptions) => Promise<ChooseImageResponse>;
export default chooseImage;
