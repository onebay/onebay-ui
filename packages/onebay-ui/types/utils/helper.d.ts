export declare const sleep: (timeout: number) => Promise<null>;
