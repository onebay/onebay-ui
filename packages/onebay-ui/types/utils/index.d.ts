import baseToString from './baseToString';
export declare function mergeStyle(style1: string | unknown, style2: string | unknown): string | unknown;
export declare function handleTouchScroll(flag: boolean): void;
export declare const isFunction: (value: unknown) => value is Function;
interface EventDetail {
    pageX: number;
    pageY: number;
    clientX: number;
    clientY: number;
    offsetX: number;
    offsetY: number;
    x: number;
    y: number;
}
export declare function getEventDetail(event: any): EventDetail;
interface ParameterErrorType {
    name?: string;
    para?: string;
    correct: string;
    wrong: unknown;
}
export declare function shouleBeObject(target: unknown): Partial<{
    res: boolean;
    msg: string;
}>;
export declare function getParameterError({ name, para, correct, wrong }: ParameterErrorType): string;
/**
 * return a uuid
 */
export declare const uuid: () => string;
export declare const isTest: () => boolean;
export { baseToString };
export declare const getScrollTop: () => number;
export interface Point {
    index: number;
    x: number;
    y: number;
}
interface TouchPoint {
    clientX: number;
    clientY: number;
    force: number;
    identifier: number;
    pageX: number;
    pageY: number;
    radiusX: number;
    radiusY: number;
    rotationAngle: number;
    screenX: number;
    screenY: number;
}
export declare const getPoint: (point: Partial<TouchPoint>) => Point;
export declare const getLen: (points: Point[]) => number;
