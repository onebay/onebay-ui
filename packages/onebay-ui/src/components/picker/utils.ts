function getTimeRange(begin: number, end: number) {
  const range: string[] = []
  for (let i = begin; i <= end; i++) {
    range.push(`${i < 10 ? '0' : ''}${i}`)
  }
  return range
}

export const hoursRange = ['20', '21', '22', '23', ...getTimeRange(0, 23), '00', '01', '02', '03']

export const minutesRange = ['56', '57', '58', '59', ...getTimeRange(0, 59), '00', '01', '02', '03']

/**
 * Check whether the passed value is legal
 */
export function verifyValue(value: number, range: unknown[]): boolean {
  if (!isNaN(+value) && value >= 0 && value < range.length) return true
  return false
}

/**
 * Check whether the passed time value is legal
 */
export function verifyTime(value: string): boolean {
  if (!/^\d{1,2}:\d{1,2}$/.test(value)) return false

  const time = value.split(':').map((num) => +num)

  if (time[0] < 0 || time[0] > 23) return false
  if (time[1] < 0 || time[1] > 59) return false

  return true
}

/**
 * compare time
 * return t1 <= t2
 */
export function compareTime(t1: string, t2: string): boolean {
  const t1List = t1.split(':').map((i) => +i)
  const t2List = t2.split(':').map((i) => +i)
  if (t1List[0] < t2List[0]) return true
  if (t1List[0] === t2List[0] && t1List[1] <= t2List[1]) return true

  return false
}

/**
 * Check the validity of the date, return the validity and date array
 */
export function verifyDate(dateStr: string): boolean | Date {
  if (!dateStr) return false
  const date = new Date(dateStr.replace(/-/g, '/'))
  return isNaN(date.getMonth()) ? false : date
}

/**
 * Get the maximum number of days in the month
 */
function getMaxDay(year: number, month: number) {
  if (month === 4 || month === 6 || month === 9 || month === 11) return 30
  if (month === 2) {
    if ((year % 4 === 0 && year % 100 !== 0) || year % 400 === 0) return 29
    else return 28
  }
  return 31
}
export type ValidValue = number | string | string[]
export function formatValue(value: number | number[] | string): ValidValue {
  let res: ValidValue
  if (Array.isArray(value)) {
    res = value.map((item) => String(item))
  } else {
    res = value
  }
  return res
}

/**
 * Get time array
 */
function getDateRange(start: number, end: number) {
  const range: number[] = []
  for (let i = start; i <= end; i++) {
    range.push(i)
  }
  return range
}

/**
 * Get an array of year intervals
 */
export function getYearRange(start: number, end: number): number[] {
  return getDateRange(start, end)
}

/**
 * Get an array of month intervals
 */
export function getMonthRange(start: Date, end: Date, year: number): number[] {
  let rangeStart = 1
  let rangeEnd = 12

  // when start year equal to the year, rangeStart should be start month
  if (start.getFullYear() === year) {
    rangeStart = start.getMonth() + 1
  }

  // when end year equal to the year, rangeEnd should be end month
  if (end.getFullYear() === year) {
    rangeEnd = end.getMonth() + 1
  }

  return getDateRange(rangeStart, rangeEnd)
}

/**
 * Get date range array
 */
export function getDayRange(start: Date, end: Date, year: number, month: number): number[] {
  let rangeStart = 1
  let rangeEnd = getMaxDay(year, month)

  if (start.getFullYear() === year && start.getMonth() + 1 === month) {
    rangeStart = start.getDate()
  }

  if (end.getFullYear() === year && end.getMonth() + 1 === month) {
    rangeEnd = end.getDate()
  }

  return getDateRange(rangeStart, rangeEnd)
}
