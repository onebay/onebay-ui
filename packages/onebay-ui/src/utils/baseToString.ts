function arrayMap(array, iteratee) {
  let index = -1
  const length = array == null ? 0 : array.length
  const result = Array(length)

  while (++index < length) {
    result[index] = iteratee(array[index], index, array)
  }
  return result
}

const INFINITY = 1 / 0

const symbolProto = Symbol ? Symbol.prototype : undefined
const symbolToString = symbolProto ? symbolProto.toString : undefined

function isSymbol(value: unknown): boolean {
  return typeof value == 'symbol'
}

export default function baseToString(value: unknown): string {
  // Exit early for strings to avoid a performance hit in some environments.
  if (typeof value == 'string') {
    return value
  }
  if (Array.isArray(value)) {
    // Recursively convert values (susceptible to call stack limits).
    return arrayMap(value, baseToString) + ''
  }
  if (isSymbol(value)) {
    return symbolToString ? symbolToString.call(value) : ''
  }
  const result = value + ''
  return result == '0' && 1 / (value as number) == -INFINITY ? '-0' : result
}
