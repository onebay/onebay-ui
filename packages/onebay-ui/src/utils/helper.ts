export const sleep = async (timeout: number): Promise<null> =>
  new Promise((resolve) => setTimeout(resolve, timeout))
