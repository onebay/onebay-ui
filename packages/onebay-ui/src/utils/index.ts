import baseToString from './baseToString'
const objectToString = (style: string | unknown): string => {
  if (style && typeof style === 'object') {
    let styleStr = ''
    Object.keys(style).forEach((key) => {
      const lowerCaseKey = key.replace(/([A-Z])/g, '-$1').toLowerCase()
      styleStr += `${lowerCaseKey}:${style[key]};`
    })
    return styleStr
  } else if (style && typeof style === 'string') {
    return style
  }
  return ''
}
export function mergeStyle(style1: string | unknown, style2: string | unknown): string | unknown {
  if (style1 && typeof style1 === 'object' && style2 && typeof style2 === 'object') {
    return Object.assign({}, style1, style2)
  }
  return objectToString(style1) + objectToString(style2)
}

let scrollTop = 0
export function handleTouchScroll(flag: boolean): void {
  if (flag) {
    scrollTop = document.documentElement.scrollTop

    // 使body脱离文档流
    document.body.classList.add('ob-frozen')

    // 把脱离文档流的body拉上去！否则页面会回到顶部！
    document.body.style.top = `${-scrollTop}px`
  } else {
    document.body.style.top = ''
    document.body.classList.remove('ob-frozen')

    document.documentElement.scrollTop = scrollTop
  }
}

// eslint-disable-next-line @typescript-eslint/ban-types
export const isFunction = (value: unknown): value is Function => typeof value === 'function'

interface EventDetail {
  pageX: number
  pageY: number
  clientX: number
  clientY: number
  offsetX: number
  offsetY: number
  x: number
  y: number
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function getEventDetail(event: any): EventDetail {
  const detail: EventDetail = {
    pageX: event.pageX,
    pageY: event.pageY,
    clientX: event.clientX,
    clientY: event.clientY,
    offsetX: event.offsetX,
    offsetY: event.offsetY,
    x: event.x,
    y: event.y
  }
  return detail
}

export function shouleBeObject(target) {
  if (target && typeof target === 'object') return { res: true }
  return {
    res: false,
    msg: getParameterError({
      correct: 'Object',
      wrong: target
    })
  }
}

export function getParameterError({ name = '', para, correct, wrong }) {
  const parameter = para ? `parameter.${para}` : 'parameter'
  const errorType = upperCaseFirstLetter(wrong === null ? 'Null' : typeof wrong)
  return `${name}:fail parameter error: ${parameter} should be ${correct} instead of ${errorType}`
}

function upperCaseFirstLetter(string: unknown) {
  if (typeof string !== 'string') return string
  string = string.replace(/^./, (match) => match.toUpperCase())
  return string
}

/**
 * return a uuid
 */
export const uuid = ((id: number) => () => `uuid-${id++}`)(1000)

export const isTest = (): boolean => process?.env?.NODE_ENV === 'test'

export { baseToString }

export const getScrollTop = (): number =>
  document.documentElement.scrollTop || document.body.scrollTop

export interface Point {
  index: number
  x: number
  y: number
}

interface TouchPoint {
  clientX: number
  clientY: number
  force: number
  identifier: number
  pageX: number
  pageY: number
  radiusX: number
  radiusY: number
  rotationAngle: number
  screenX: number
  screenY: number
}

export const getPoint = (point: Partial<TouchPoint>): Point => {
  return {
    index: point.identifier,
    x: point.clientX,
    y: point.clientY
  }
}

export const getLen = (points: Point[]): number => {
  const len = Math.sqrt(
    Math.pow(points[0].x - points[1].x, 2) + Math.pow(points[0].y - points[1].y, 2)
  )
  return len
}
