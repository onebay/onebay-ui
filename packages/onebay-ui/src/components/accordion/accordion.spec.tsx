import { mount } from '@vue/test-utils'
import Accordion from './index'
import { sleep } from '../../utils/helper'

const factory = (values = {}, slots = { default: ['按钮'] }) => {
  return mount(Accordion, {
    slots,
    props: { ...values }
  })
}

describe('Accordion Snap', () => {
  it('render initial Accordion', () => {
    const wrapper = mount(Accordion)
    expect(wrapper.element).toMatchSnapshot()
  })

  it('render Accordion -- props title', () => {
    const wrapper = mount(<Accordion title="title" />)
    expect(wrapper.element).toMatchSnapshot()
  })

  it('render Accordion -- props open', () => {
    const wrapper = mount(<Accordion open={true} />)
    expect(wrapper.element).toMatchSnapshot()
  })

  it('render Accordion -- props isAnimation', async () => {
    const wrapper = mount(<Accordion isAnimation={false} />)
    await wrapper.setProps({ open: true })
    await sleep(0)
    expect(wrapper.element).toMatchSnapshot()
  })

  it('render Accordion -- props icon', () => {
    const wrapper = factory(
      {
        icon: { value: 'chevron-down', color: 'red' }
      },
      {
        default: ['<div></div>']
      }
    )
    expect(wrapper.element).toMatchSnapshot()
  })

  it('render Accordion -- props icon prefixClass', () => {
    const wrapper = factory(
      {
        icon: { prefixClass: 'prefixClass', value: 'star', color: 'red' }
      },
      {
        default: ['<div></div>']
      }
    )
    expect(wrapper.element).toMatchSnapshot()
  })

  it('render Accordion -- props note', () => {
    const wrapper = factory(
      {
        note: 'note'
      },
      {
        default: ['<div></div>']
      }
    )
    expect(wrapper.element).toMatchSnapshot()
  })

  it('render Accordion -- props note', () => {
    const wrapper = factory(
      {
        note: 'note'
      },
      {
        default: ['<div></div>']
      }
    )
    expect(wrapper.element).toMatchSnapshot()
  })

  it('test change props open', async () => {
    const wrapper = factory()
    await wrapper.setProps({ open: true })
    expect(wrapper.vm.open).toBeTruthy()
    expect(wrapper.element).toMatchSnapshot()
    await sleep(700)
  })
})

describe('Accordion Event', () => {
  it('toggle accordion', () => {
    const wrapper = factory()
    wrapper.find('.ob-accordion__header').trigger('click')
    expect(wrapper.emitted()).toHaveProperty('click')
  })

  it('toggle accordion animation is not completed', async () => {
    const wrapper = factory()
    await wrapper.setProps({ open: true })
    wrapper.find('.ob-accordion__header').trigger('click')
    expect(wrapper.emitted()).not.toHaveProperty('click')
  })
})
