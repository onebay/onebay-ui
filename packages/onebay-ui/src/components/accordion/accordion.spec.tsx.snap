// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion Snap render Accordion -- props icon 1`] = `
<div
  class="ob-accordion"
>
  <div
    class="ob-accordion__header"
  >
    <div
      class="ob-icon ob-icon-chevron-down ob-accordion__icon"
      style="color: red; font-size: undefinedpx;"
    />
    <div
      class="ob-accordion__info"
    >
      <div
        class="ob-accordion__info__title"
      >
        
      </div>
      <div
        class="ob-accordion__info__note"
      >
        
      </div>
    </div>
    <div
      class="ob-accordion__arrow"
    >
      <div
        class="ob-icon ob-icon-chevron-down"
      />
    </div>
  </div>
  <div
    class="ob-accordion__content ob-accordion__content--inactive"
  >
    <div
      class="ob-accordion__body"
    >
      
      &lt;div&gt;&lt;/div&gt;
      
    </div>
  </div>
</div>
`;

exports[`Accordion Snap render Accordion -- props icon prefixClass 1`] = `
<div
  class="ob-accordion"
>
  <div
    class="ob-accordion__header"
  >
    <div
      class="prefixClass prefixClass-star ob-accordion__icon"
      style="color: red; font-size: undefinedpx;"
    />
    <div
      class="ob-accordion__info"
    >
      <div
        class="ob-accordion__info__title"
      >
        
      </div>
      <div
        class="ob-accordion__info__note"
      >
        
      </div>
    </div>
    <div
      class="ob-accordion__arrow"
    >
      <div
        class="ob-icon ob-icon-chevron-down"
      />
    </div>
  </div>
  <div
    class="ob-accordion__content ob-accordion__content--inactive"
  >
    <div
      class="ob-accordion__body"
    >
      
      &lt;div&gt;&lt;/div&gt;
      
    </div>
  </div>
</div>
`;

exports[`Accordion Snap render Accordion -- props isAnimation 1`] = `
<div
  class="ob-accordion"
>
  <div
    class="ob-accordion__header"
  >
    
    <div
      class="ob-accordion__info"
    >
      <div
        class="ob-accordion__info__title"
      >
        
      </div>
      <div
        class="ob-accordion__info__note"
      >
        
      </div>
    </div>
    <div
      class="ob-accordion__arrow ob-accordion__arrow--folded"
    >
      <div
        class="ob-icon ob-icon-chevron-down"
      />
    </div>
  </div>
  <div
    class="ob-accordion__content"
  >
    <div
      class="ob-accordion__body"
    >
      <!---->
    </div>
  </div>
</div>
`;

exports[`Accordion Snap render Accordion -- props note 1`] = `
<div
  class="ob-accordion"
>
  <div
    class="ob-accordion__header"
  >
    
    <div
      class="ob-accordion__info"
    >
      <div
        class="ob-accordion__info__title"
      >
        
      </div>
      <div
        class="ob-accordion__info__note"
      >
        note
      </div>
    </div>
    <div
      class="ob-accordion__arrow"
    >
      <div
        class="ob-icon ob-icon-chevron-down"
      />
    </div>
  </div>
  <div
    class="ob-accordion__content ob-accordion__content--inactive"
  >
    <div
      class="ob-accordion__body"
    >
      
      &lt;div&gt;&lt;/div&gt;
      
    </div>
  </div>
</div>
`;

exports[`Accordion Snap render Accordion -- props note 2`] = `
<div
  class="ob-accordion"
>
  <div
    class="ob-accordion__header"
  >
    
    <div
      class="ob-accordion__info"
    >
      <div
        class="ob-accordion__info__title"
      >
        
      </div>
      <div
        class="ob-accordion__info__note"
      >
        note
      </div>
    </div>
    <div
      class="ob-accordion__arrow"
    >
      <div
        class="ob-icon ob-icon-chevron-down"
      />
    </div>
  </div>
  <div
    class="ob-accordion__content ob-accordion__content--inactive"
  >
    <div
      class="ob-accordion__body"
    >
      
      &lt;div&gt;&lt;/div&gt;
      
    </div>
  </div>
</div>
`;

exports[`Accordion Snap render Accordion -- props open 1`] = `
<div
  class="ob-accordion"
>
  <div
    class="ob-accordion__header"
  >
    
    <div
      class="ob-accordion__info"
    >
      <div
        class="ob-accordion__info__title"
      >
        
      </div>
      <div
        class="ob-accordion__info__note"
      >
        
      </div>
    </div>
    <div
      class="ob-accordion__arrow ob-accordion__arrow--folded"
    >
      <div
        class="ob-icon ob-icon-chevron-down"
      />
    </div>
  </div>
  <div
    class="ob-accordion__content"
  >
    <div
      class="ob-accordion__body"
    >
      <!---->
    </div>
  </div>
</div>
`;

exports[`Accordion Snap render Accordion -- props title 1`] = `
<div
  class="ob-accordion"
>
  <div
    class="ob-accordion__header"
  >
    
    <div
      class="ob-accordion__info"
    >
      <div
        class="ob-accordion__info__title"
      >
        title
      </div>
      <div
        class="ob-accordion__info__note"
      >
        
      </div>
    </div>
    <div
      class="ob-accordion__arrow"
    >
      <div
        class="ob-icon ob-icon-chevron-down"
      />
    </div>
  </div>
  <div
    class="ob-accordion__content ob-accordion__content--inactive"
  >
    <div
      class="ob-accordion__body"
    >
      <!---->
    </div>
  </div>
</div>
`;

exports[`Accordion Snap render initial Accordion 1`] = `
<div
  class="ob-accordion"
>
  <div
    class="ob-accordion__header"
  >
    
    <div
      class="ob-accordion__info"
    >
      <div
        class="ob-accordion__info__title"
      >
        
      </div>
      <div
        class="ob-accordion__info__note"
      >
        
      </div>
    </div>
    <div
      class="ob-accordion__arrow"
    >
      <div
        class="ob-icon ob-icon-chevron-down"
      />
    </div>
  </div>
  <div
    class="ob-accordion__content ob-accordion__content--inactive"
  >
    <div
      class="ob-accordion__body"
    >
      <!---->
    </div>
  </div>
</div>
`;

exports[`Accordion Snap test change props open 1`] = `
<div
  class="ob-accordion"
>
  <div
    class="ob-accordion__header"
  >
    
    <div
      class="ob-accordion__info"
    >
      <div
        class="ob-accordion__info__title"
      >
        
      </div>
      <div
        class="ob-accordion__info__note"
      >
        
      </div>
    </div>
    <div
      class="ob-accordion__arrow ob-accordion__arrow--folded"
    >
      <div
        class="ob-icon ob-icon-chevron-down"
      />
    </div>
  </div>
  <div
    class="ob-accordion__content"
    style="height: 0px;"
  >
    <div
      class="ob-accordion__body"
    >
      
      按钮
      
    </div>
  </div>
</div>
`;
